torch/csrc/Device.cpp

```diff
@@ -0,0 +1,217 @@
+#include "Device.h"
+
+#include <cstring>
+#include <structmember.h>
+#include <sstream>
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/utils/object_ptr.h"
+#include "torch/csrc/utils/python_arg_parser.h"
+#include "torch/csrc/utils/python_strings.h"
+
+PyObject *THPDevice_New(torch::DeviceType device_type, int64_t device_index, bool is_default)
+{
+  auto type = (PyTypeObject*)&THPDeviceType;
+  auto self = THPObjectPtr{type->tp_alloc(type, 0)};
+  if (!self) throw python_error();
+  auto self_ = reinterpret_cast<THPDevice*>(self.get());
+  self_->device_type = device_type;
+  self_->device_index = device_index;
+  self_->is_default = is_default;
+  return self.release();
+}
+
+static const char* cuda_str = "cuda";
+static const char* cpu_str = "cpu";
+
+static inline const char* deviceTypeString(torch::DeviceType device_type) {
+  switch (device_type) {
+    case torch::DeviceType::CUDA:
+      return cuda_str;
+    case torch::DeviceType::CPU:
+      return cpu_str;
+    default:
+      throw std::runtime_error("unexpected device type");
+  }
+}
+
+PyObject *THPDevice_repr(THPDevice *self)
+{
+  std::ostringstream oss;
+  oss << "Device(device_type=\'" << deviceTypeString(self->device_type) << "\'";
+  if (!self->is_default) {
+    oss << ", device_index=" << self->device_index;
+  }
+  oss << ")";
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDevice_str(THPDevice*self)
+{
+  std::ostringstream oss;
+  if (!self->is_default) {
+    oss << deviceTypeString(self->device_type) << ":" << self->device_index;
+  } else {
+    oss << deviceTypeString(self->device_type);
+  }
+  return THPUtils_packString(oss.str().c_str());
+}
+
+PyObject *THPDevice_pynew(PyTypeObject *type, PyObject *args, PyObject *kwargs)
+{
+  HANDLE_TH_ERRORS
+  static torch::PythonArgParser parser({
+    "Device(Device device)",
+    "Device(String device_type, int64_t? device_index=-1)"
+  });
+  torch::ParsedArgs<2> parsed_args;
+  auto r = parser.parse(args, kwargs, parsed_args);
+  if (r.idx == 0) {
+    auto device = r.device(0);
+    return THPDevice_New(device.device_type, device.device_index, device.is_default);
+  } else if (r.idx == 1) {
+    auto as_device = r.device(0);  // this works, because device can take strings
+    auto device_type = r.string(0);
+    if (!as_device.is_default) {
+      throw std::runtime_error("device_type must not include index because index is passed explicitly " + device_type);
```

Comment:
Is it? It looks like the signature allows passing in only the string

Reply:
The message is confusing; it means the device_type string parameter for the second overload must only be "cuda" or "cpu" because the index is a separate parameter there.